server/dns_server.ml

```diff
@@ -40,11 +41,12 @@ module Authentication = struct
     (* the name of a key is primaryip.secondaryip._transfer.zone
        e.g. 192.168.42.2.192.168.42.1._transfer.mirage
        alternative: <whatever>.primaryip._transfer.zone *)
-    let arr = Domain_name.to_array name in
-    let transfer = operation_to_string `Transfer in
-    try
-      let rec go idx = if Array.get arr idx = transfer then idx else go (succ idx) in
-      let zone_idx = go 0 in
+    let is_transfer = Domain_name.equal_label (operation_to_string `Transfer) in
+    match Domain_name.find_label ~rev:true name is_transfer with
+    | None -> None
+    | Some idx ->
+      let arr = Domain_name.to_array name in
+      let zone_idx = Array.length arr - succ idx in
       let zone = Domain_name.of_array (Array.sub arr 0 zone_idx) in
```

Comment:
now that we have `drop_label` I think this can be solved with something like (or `~rev:true` if we want the other side?)
```ocaml
let zone = Domain_name.drop_label_exn ~amount:(succ idx) d in
```

Reply:
yes, but in the following code there's still some functionality missing from domain-name, namely "give me a subdomain starting at index y and length l" -- that's why i got lazy and didn't bother to convert this fragment to not use Domain_name.to_array/of_array yet...